### PR TITLE
New parallel optimizer

### PR DESCRIPTION
**Pull request by wzhongyuan**
## What changes were proposed in this pull request?

1) Add a new Optimizer `ParallelOptimizer` to do parameter sync asynchronously
2) Compatible with original DistriOptimizer (transparent)
3) fully integrated with MKL-DNN
## How was this patch tested?
Resnet-50 training, lenet training

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ConcatTable.scala [hunk 1]
```diff
@@ -58,6 +58,7 @@ class ConcatTable extends MklDnnContainer {
     var i = 0
     while (i < modules.length) {
       modules(i).accGradParameters(input, gradOutput.toTable(i + 1))
+      modules(i).asyncGradient
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ConcatTable.scala]: So what will happen if we use `DistriOptimizer` ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Sequential.scala [hunk 2]
```diff
@@ -139,6 +139,7 @@ class Sequential extends MklDnnContainer {
         modules(i - 1).output
       )
       currentModule.accGradParameters(curInput, lastGradInput)
+      currentModule.asyncGradient
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Sequential.scala]: So what will happen if we use `DistriOptimizer` ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Sequential.scala [hunk 3]
```diff
@@ -148,6 +149,7 @@ class Sequential extends MklDnnContainer {
     }
 
     modules(i).accGradParameters(input, lastGradInput)
+    modules(i).asyncGradient
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Sequential.scala]: So what will happen if we use `DistriOptimizer` ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala [hunk 4]
```diff
@@ -0,0 +1,1053 @@
...
+   * @param validationTrigger validation trigger
+   * @param validationDataSet validation dataset
+   * @param validationMethods validation methods
+   * @param cacheTrigger cache trigger
+   * @param cachePath cache path
+   * @param trainSummary train summary
+   * @param validationSummary validation summary
+   * @param isOverWrite if overwrite the checkpoint
+   */
+  private[optim] def optimize[T: ClassTag](
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: Should we add some unit tests for this important method, including local, spark local, spark distributed?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 5]
```diff
@@ -1115,5 +1124,60 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
    * JVM GC can't release them reliably.
    */
   def release(): Unit = {}
+
+
+  /**
+   * parameter synchronizer for gradient synchronization
+   */
+  var _parameterSynchronizer: DistriParameterSynchronizer[T] = null
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: private?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 6]
```diff
@@ -1115,5 +1124,60 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+
+  /**
+   * set parameter synchronizer
+   * @param parameterSynchronizer parameter synchronizer
+   */
+  def setParameterSynchronizer(parameterSynchronizer: DistriParameterSynchronizer[T]): Unit = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: private?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 7]
```diff
@@ -1115,5 +1124,60 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+    _parameterSynchronizer = parameterSynchronizer
+  }
+
+
+  /**
+   * get parameter synchronizer
+   * @return parameter synchronizer
+   */
+  def getParameterSynchronizer(): DistriParameterSynchronizer[T] = _parameterSynchronizer
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: private?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 8]
```diff
@@ -1115,5 +1124,60 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+
+
+  var _optimMethod: OptimMethod[T] = null
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: private?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 9]
```diff
@@ -71,7 +72,8 @@ object DistriOptimizer {
     localStates: Array[Table],
     var moduleTimeList: Array[Long] = null,
     localMethods: Array[Option[Array[ValidationMethod[T]]]],
-    optimMethods: Map[String, OptimMethod[T]]
+    optimMethods: Map[String, OptimMethod[T]],
+    parameterSynchronizer: DistriParameterSynchronizer[T] = null
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: why we cache this?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala (hunk 4 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: Added, this is only for spark not for local mode

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 5 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: refined

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 6 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: refined

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 7 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: refined

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 8 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: refined

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala (hunk 9 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: We want to release all the resources at last and also we need to reference this object to get initially set partition ID

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala [hunk 10]
```diff
@@ -0,0 +1,1053 @@
...
+  }
+
+  /**
+   * Save train summaries.
+   *
+   * @param trainSummary train logger
+   * @param models cached models
+   * @param driverState driver state
+   */
+  private def saveSummary[T: ClassTag](
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: can we share this method with the original optimizer?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala [hunk 11]
```diff
@@ -0,0 +1,1053 @@
...
+   * @param criterion loss function
+   * @param state state table
+   * @param nodeNumber node number
+   * @param coresPerNode cores per node
+   * @param checkSingleton if checkSingleton
+   * @param validationMethods validation methods
+   * @param optimMethod optimization method
+   * @return cached models
+   */
+  private def initThreadModels[T: ClassTag](
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: same with above

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala [hunk 12]
```diff
@@ -0,0 +1,1053 @@
...
+      val (numFinishedModelUpdates, localLossSum, localRecordsNum) = dataRDD
+        .zipPartitions(models, preservesPartitioning = true) { (data, modelIter) => {
+          var count = 0
+          var finishedThreadSize = 0
+          val cached = modelIter.next()
+          val miniBatchBuffer = new Array[MiniBatch[T]](_subModelNumber)
+          while (count < iterationPerTime) {
+            val syWStart = System.nanoTime()
+            val batch = data.next()
+            val stackSize = batch.size() / _subModelNumber
```
[the remark this answers is not in the document]
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: Makes sense, I'll refine it

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala (hunk 10 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: yep, makes sense

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala (hunk 11 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: this method behaviours quite different from the original one (setting synchronizer, etc), I'll check if we can do some consolidation or just leave it

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala [hunk 13]
```diff
@@ -0,0 +1,1053 @@
...
+   */
+  def clearState() : Unit = {
+    // Reset the singleton flag, so other optimizers can run
+    models.mapPartitions(iter => {
+      Engine.resetSingletonFlag()
+      iter
+    }).count()
+  }
+
+  private def endEpoch(): Unit = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: can this be shared?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala [hunk 14]
```diff
@@ -0,0 +1,1053 @@
...
+    this._priorities = priorities
+  }
+
+  /**
+   * Clean some internal states, so this or other optimizers can run optimize again
+   *
+   * This method will be called at the end of optimize. You need not call it if optimize succeed.
+   * If the optimize fails, you may call it before next optimize.
+   */
+  def clearState() : Unit = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: can this be shared with the old optimizer

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala [hunk 15]
```diff
@@ -0,0 +1,1053 @@
...
+    optimMethods.foreach { case (moduleName, optimMethod) =>
+      val records = optimMethod.state.get[Int]("recordsProcessedThisEpoch")
+      if (records.isDefined && records.get != 0) {
+        optimMethod.state("epoch") = optimMethod.state[Int]("epoch") + 1
+        optimMethod.state("recordsProcessedThisEpoch") = 0
+      }
+    }
+  }
+
+  override def setTrainData(sampleRDD: RDD[Sample[T]],
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: same with the above

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala [hunk 16]
```diff
@@ -0,0 +1,1053 @@
...
+    this.dataset = (DataSet.rdd(sampleRDD) ->
+      SampleToMiniBatch(miniBatch, batchSize, None))
+      .asInstanceOf[DistributedDataSet[MiniBatch[T]]]
+    // if current epoch is not finished, we will end the
+    // current epoch and start a new epoch when optimize is called
+    endEpoch()
+    this
+  }
+
+  override def setTrainData(sampleRDD: RDD[Sample[T]],
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: same

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala [hunk 17]
```diff
@@ -0,0 +1,1053 @@
...
+      SampleToMiniBatch(batchSize, _featurePaddingParam, _labelPaddingParam))
+      .asInstanceOf[DistributedDataSet[MiniBatch[T]]]
+    // if current epoch is not finished, we will end the
+    // current epoch and start a new epoch when optimize is called
+    endEpoch()
+    this
+  }
+
+
+  override def prepareInput(): Unit = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: same

**Comment by wzhongyuan** [comment]
@yiheng  refined per your comments, please help review, thanks

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala [hunk 18]
```diff
@@ -0,0 +1,861 @@
...
+        optimMethod.state.update("recordsProcessedThisEpoch", 0)
+      }
+    }
+
+    val _subModelNumber = Engine.getEngineType() match {
+      case MklBlas => coresPerNode
+      case MklDnn => 1
+    }
+
+    require(_subModelNumber == 1, "currently only single model supported especially for mkldnn")
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: why not just check the engineType==dnn...

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala [hunk 19]
```diff
@@ -0,0 +1,861 @@
...
+      subModules.foreach(sub => {
+        res ++= getExecutionOrder(sub)
+      })
+    } else {
+      res += module
+    }
+    res
+  }
+
+  private def setDistriPartitionsynchronizer[T: ClassTag](model: Module[T],
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: setDistriPartitionSynchronizer ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala [hunk 20]
```diff
@@ -0,0 +1,861 @@
...
+    val priorities = new mutable.HashMap[String, Int]
+    val orders = getExecutionOrder(this._model)
+    val len = orders.size
+    orders.zipWithIndex.foreach(order => {
+      priorities.put(order._1.getName, len - order._2)
+    })
+    priorities
+  }
+
+  private def getExecutionOrder(module : Module[T]): ArrayBuffer[Module[T]] = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala]: duplicate method?

**Comment by wzhongyuan** [comment]
PR validation passed, merging it now

**Pull request merged**